### PR TITLE
Add GitHub action to format and lint code

### PR DESCRIPTION
**Pull request by dyastremsky**
This pull request adds pre-commit hooks and updates the contributing guidelines to reflect that we are using them for validating commits in a pull request. It also includes running the pre-commit hooks on these files, plus making fixes based on them (e.g. making sure all `.py` and `.sh` executables are set to be executable and have shebangs, spelling corrections, etc.).

This pull request adds pre-commit  hooks to run the following:
- Black for formatting Python
- Flake8 for linting Python for conformance to PEP8, PyFlakes, and circular complexity (max-line-length set to match Black's 88 default/recommended character limit; this and other settings from Black's documentation [here](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#line-length) to avoid enforcement of limit when unnecessary).
- isort for deterministically sorting and organizing Python imports
- Clang for formatting C++, C, Java, TextProto, Proto, and CUDA
- Codespell for finding common spelling mistakes 

In addition, this adds these native pre-commit hooks, details [here](https://pre-commit.com/hooks.html):
- check-case-conflict
- check-executables-have-shebangs
- check-merge-conflict
- check-json
- check-toml
- check-yaml
- check-shebang-scripts-are-executable
- end-of-file-fixer
- mixed-line-ending
- requirements-txt-fixer
- trailing-whitespace

To run these locally, you can go into the repo and run `pip install pre-commit` and `pre-commit install`. After that has been done once, you just need to run `pre-commit run --all-files` any time you want to apply them. Once installed, it should run automatically on commit for files included in a commit. To correct spelling errors found by codespell when possible, you can install codespell (`pip install codespell` and potentially `pip install --upgrade codespell` to use the toml file in the directory) and call `codespell -w <path>`.

This pull request also adds a GitHub action so that these are run for every pull request and push to main. The core changes are in pre-commit.yml, pre-commit-config.yaml, and pyproject.toml. The rest of the changes are the effect of applying the action to the current repo.

### CODE REVIEW

File: build.py
Comment (nnshah1) [inline build.py:77]:
```suggestion
        "py310_23.1.0-1", #Conda version
```

File: build.py
Comment (nnshah1) [inline build.py:78]:
```suggestion
    )  # Conda version.
```
```suggestion
    )  
```

File: deploy/aws/README.md
Comment (oandreeva-nv) [inline deploy/aws/README.md:42]: Should we keep the length below 80 characters? Can we set this with hooks?

File: docs/conf.py
Comment (oandreeva-nv) [inline docs/conf.py:74]: Do we need to keep it?

File: docs/examples/jetson/concurrency_and_dynamic_batching/README.md
Comment (oandreeva-nv) [inline docs/examples/jetson/concurrency_and_dynamic_batching/README.md:1]: Please, update copyright

File: docs/getting_started/quickstart.md
Comment (oandreeva-nv) [inline docs/getting_started/quickstart.md:1]: Please, update copyright

File: docs/protocol/extension_model_configuration.md
Comment (oandreeva-nv) [inline docs/protocol/extension_model_configuration.md:1]: Copyright

File: docs/protocol/extension_logging.md
Comment (oandreeva-nv) [inline docs/protocol/extension_logging.md:1]: copyright

File: docs/examples/jetson/concurrency_and_dynamic_batching/Makefile
Comment (oandreeva-nv) [inline docs/examples/jetson/concurrency_and_dynamic_batching/Makefile:1]: ditto

File: docs/examples/README.md
Comment (oandreeva-nv) [inline docs/examples/README.md:1]: ditto

File: docs/_static/custom.css
Comment (oandreeva-nv) [inline docs/_static/custom.css:1]: copyright

File: docs/Makefile
Comment (oandreeva-nv) [inline docs/Makefile:1]: copyright

File: docker/sagemaker/serve
Comment (oandreeva-nv) [inline docker/sagemaker/serve:1]: copyright

File: deploy/mlflow-triton-plugin/mlflow_triton/deployments.py
Comment (oandreeva-nv) [inline deploy/mlflow-triton-plugin/mlflow_triton/deployments.py:34]: I'm curious, why this wasn't grouped with other `from ..`

File: deploy/gke-marketplace-app/server-deployer/chart/triton/templates/service.yaml
Comment (oandreeva-nv) [inline deploy/gke-marketplace-app/server-deployer/chart/triton/templates/service.yaml:1]: copyright

File: deploy/gke-marketplace-app/server-deployer/chart/triton/templates/deployment.yaml
Comment (oandreeva-nv) [inline deploy/gke-marketplace-app/server-deployer/chart/triton/templates/deployment.yaml:1]: copyright

File: deploy/gke-marketplace-app/server-deployer/chart/triton/templates/application.yaml
Comment (oandreeva-nv) [inline deploy/gke-marketplace-app/server-deployer/chart/triton/templates/application.yaml:1]: copyright

File: deploy/gke-marketplace-app/benchmark/model-store/bert_distill_tf_gpu/config.pbtxt
Comment (oandreeva-nv) [inline deploy/gke-marketplace-app/benchmark/model-store/bert_distill_tf_gpu/config.pbtxt:1]: copyright

File: deploy/gke-marketplace-app/benchmark/model-store/bert_distill_tf_cpu/config.pbtxt
Comment (oandreeva-nv) [inline deploy/gke-marketplace-app/benchmark/model-store/bert_distill_tf_cpu/config.pbtxt:1]: copyright

File: deploy/gke-marketplace-app/benchmark/model-store/bert_base_trt_gpu/config.pbtxt
Comment (oandreeva-nv) [inline deploy/gke-marketplace-app/benchmark/model-store/bert_base_trt_gpu/config.pbtxt:1]: copyright

File: deploy/gke-marketplace-app/benchmark/model-store/bert_base_tf_gpu/config.pbtxt
Comment (oandreeva-nv) [inline deploy/gke-marketplace-app/benchmark/model-store/bert_base_tf_gpu/config.pbtxt:1]: copyright

File: deploy/gke-marketplace-app/benchmark/README.md
Comment (oandreeva-nv) [inline deploy/gke-marketplace-app/benchmark/README.md:1]: copyright

File: deploy/gke-marketplace-app/README.md
Comment (oandreeva-nv) [inline deploy/gke-marketplace-app/README.md:170]: Again, do we want to adjust the length of text to be below 80 chars in this PR?

File: deploy/gcp/README.md
Comment (oandreeva-nv) [inline deploy/gcp/README.md:1]: copyright

File: deploy/fleetcommand/README.md
Comment (oandreeva-nv) [inline deploy/fleetcommand/README.md:1]: copyright

File: deploy/aws/README.md
Comment (oandreeva-nv) [inline deploy/aws/README.md:1]: copyright

File: deploy/mlflow-triton-plugin/mlflow_triton/deployments.py
Comment (dyastremsky) [reply deploy/mlflow-triton-plugin/mlflow_triton/deployments.py:34]: It looks like iSort is putting these imports into a different section than others. According to the [documentation](https://pycqa.github.io/isort/docs/configuration/custom_sections_and_ordering.html), the default section ordering is: `FUTURE,STDLIB,THIRDPARTY,FIRSTPARTY,LOCALFOLDER`. I think the important part is that it's deterministic, so we don't have to worry about collisions or import order.

File: deploy/alibaba-cloud/README.md
Comment (oandreeva-nv) [inline deploy/alibaba-cloud/README.md:1]: copyright

File: deploy/mlflow-triton-plugin/mlflow_triton/deployments.py
Comment (oandreeva-nv) [reply deploy/mlflow-triton-plugin/mlflow_triton/deployments.py:34]: Thanks for explanation!

File: deploy/aws/README.md
Comment (dyastremsky) [reply deploy/aws/README.md:42]: Great idea. We had not previously enforced this, but I updated `.clang-format` to do so and will copy the change to the other repos.

File: docs/conf.py
Comment (dyastremsky) [reply docs/conf.py:74]: Good find. Tagging @dzier who made the original commit and likely knows more about this line. 

However, I'd like to avoid making changes wherever possible that are not related to formatting/linting in this PR given it already touches so many files.

File: deploy/alibaba-cloud/README.md
Comment (dyastremsky) [reply deploy/alibaba-cloud/README.md:1]: Thanks, updated all copyrights.

File: deploy/aws/README.md
Comment (dyastremsky) [reply deploy/aws/README.md:42]: I added this for the listed filetypes for clang-format. I did not enforce this for text or markdown files, as when I went to try running it, it looked like it broke things we don't want (e.g. the bug template's formatting).

One option could be creating a separate style configuration for text files and pointing a second clang-format pre-commit hook to that file, which enforces line length. We would do it going forward rather than fixing current long lines. That would avoid breaking anything and force us to think creatively to keep lines <80 characters in the future. I'd like to keep that out of this change though, as it's inconsistent with the way the other hooks are being run as part of the change and it increases the scope. I'll create a JIRA ticket to consider this change.

File: docker/entrypoint.d/56-network-driver-version-check.sh
Comment (oandreeva-nv) [inline docker/entrypoint.d/56-network-driver-version-check.sh:1]: Could you please clarify to me why we need `entrypoint.d` files? And are we keeping them empty, or we are adding `#!/bin/bash` to them? I think we need to be consistent

File: docs/customization_guide/inference_protocols.md
Comment (oandreeva-nv) [inline docs/customization_guide/inference_protocols.md:152]: I am not sure I understand both versions old and new.

File: docs/examples/jetson/concurrency_and_dynamic_batching/Makefile
Comment (oandreeva-nv) [inline docs/examples/jetson/concurrency_and_dynamic_batching/Makefile:1]: should it be 2021-2023?

File: docker/entrypoint.d/56-network-driver-version-check.sh
Comment (dyastremsky) [reply docker/entrypoint.d/56-network-driver-version-check.sh:1]: I don't know, this would be a better question for @cliffwoolley who made the change. This pull request is centered around formatting exclusively, so executable files should have shebangs. That should not change its core functionality.

I appreciate you reviewing this PR. It might make sense to pause work on this pull request. We're discussing doing a rollout in waves. Since this is the biggest code change, it'd probably be in the last wave and given all the rebasing it'll need, it'll probably need a fresh review then (sorry!). The alternative is trying to get it in now/first, but I think we'd be hard-pressed to make that happen without dedicated reviewer time (e.g. getting you and any other reviewers tickets to review this, given its size).

File: docs/examples/jetson/concurrency_and_dynamic_batching/Makefile
Comment (dyastremsky) [reply docs/examples/jetson/concurrency_and_dynamic_batching/Makefile]: Good catch, fixed.

File: docs/protocol/extension_trace.md
Comment (oandreeva-nv) [inline docs/protocol/extension_trace.md:81]: I think it is ok to accept this change for now. I can make a note to rephrase this in a follow-up PR

File: docs/customization_guide/inference_protocols.md
Comment (dyastremsky) [reply docs/customization_guide/inference_protocols.md:152]: I was only correcting the spelling, @GuanLuo might be able to clarify what this means.

File: docs/user_guide/metrics.md
Comment (oandreeva-nv) [inline docs/user_guide/metrics.md:54]: at which?

File: docs/user_guide/trace.md
Comment (oandreeva-nv) [inline docs/user_guide/trace.md:84]: Same here. Let me know if this needs to be rephrased

File: docs/user_guide/metrics.md
Comment (dyastremsky) [reply docs/user_guide/metrics.md]: Nice catch, fixed.

File: docs/user_guide/trace.md
Comment (dyastremsky) [reply docs/user_guide/trace.md:84]: Nice catch! Yeah, that could be good.

File: docker/entrypoint.d/56-network-driver-version-check.sh
Comment (oandreeva-nv) [reply docker/entrypoint.d/56-network-driver-version-check.sh:1]: No worries. I thought initial screen could speed up future work, since found issues would still need to be fixed. I'll pause for now.

Comment (cliffwoolley) [reply docker/entrypoint.d/56-network-driver-version-check.sh:1]: The reason for these to be here is to override the files of the same name in our base image.  Please don't change them.  They are sourced rather than executed directly so the shebang is irrelevant.

Comment (dyastremsky) [reply docker/entrypoint.d/56-network-driver-version-check.sh]: Thanks for clarifying, Cliff. And thank you Olga for inquiring. Reverted these changes.

Comment (dyastremsky) [reply docker/entrypoint.d/56-network-driver-version-check.sh]: Olga, feel free to resume reviewing if you find time. No worries if you can't, especially given the size of this PR. We got most of the other repos in and I realized that there's a pretty easy and safe way to rebase this (wherever there are merge conflicts, we can just use the current server repo file and re-run the hooks), confirmed by CI continuing to pass. So figure we can work to get this in.

Also, it might be worthwhile to create a JIRA ticket to account for the time you're working on this.

File: deploy/fleetcommand/README.md
Comment (krishung5) [inline deploy/fleetcommand/README.md:90]: Need to update the copyright of this file.

File: docs/protocol/extension_sequence.md
Comment (krishung5) [inline docs/protocol/extension_sequence.md:53]: Copyright

File: docs/protocol/extension_statistics.md
Comment (krishung5) [inline docs/protocol/extension_statistics.md:230]: copyright

File: deploy/fleetcommand/README.md
Comment (dyastremsky) [reply deploy/fleetcommand/README.md:90]: Thanks, updated these!

File: README.md
Comment (rmccorm4) [inline README.md:209]:
```suggestion
- Create [decoupled backends and models](docs/user_guide/decoupled_models.md) that can send
```

File: build.py
Comment (rmccorm4) [inline build.py:78]: This still looks wrong, I wonder if it's not possible to have a comment on the last item of the tuple in the rules?

Comment (rmccorm4) [reply build.py:78]: Longer term improvement that could fix this would be to change the tuple to a nested dict where the keys reflect the versions. I think that will clarify the usage throughout the code as well.

Instead of 
```python
# Access ORT version
TRITON_VERSION_MAP[version][2]
```
it could be:
```python
TRITON_VERSION_MAP = {
    "2.37.0dev": {
        "triton_container": "23.08dev",
        "upstream_container": "23.06",  
        "onnxruntime": "1.15.0", 
    # ...
}

TRITON_VERSION_MAP[version]["onnxruntime"]
```

Comment (dyastremsky) [reply build.py]: 100% agree with that recommendation! Really like the improvement, and I know @nnshah1 and @nv-kmcgill53 are currently looking into build.py improvements, so this could be a great direction.

This wasn't a pre-commit hook issue so much as the easiest/safest way to rebase being to use the latest file (in this case `build.py`) then re-run pre-commit on it. I try to integrate any previously-applied changes/feedback in, but I must have missed this one. Thanks for catching this, Ryan!

Comment (rmccorm4) [reply build.py]: LMK if there are any tickets on build.py improvements and I can link this comment or something.

File: build.py
Comment (rmccorm4) [inline build.py:2489]: Yet another separate follow-up idea since the PR is huge, but I think standardizing to f-strings would be nice and remove some extra lines of code throughout. Looks like there's a tool to automate replacement: https://github.com/ikamensh/flynt that may help.

File: docker/cpu_only/entrypoint.d/12-banner.sh
Comment (rmccorm4) [inline docker/cpu_only/entrypoint.d/12-banner.sh:1]: Do you know why these are being removed?

File: docs/README.md
Comment (rmccorm4) [inline docs/README.md:192]:
```suggestion
- **Backends**: Triton supports a wide variety of frameworks used to run models. Users can extend this functionality by creating custom backends.
```

File: build.py
Comment (nv-kmcgill53) [reply build.py]: I was going to use today to write up that proposal but got bogged down on other priorities.

File: src/grpc/stream_infer_handler.cc
Comment (rmccorm4) [inline src/grpc/stream_infer_handler.cc:221]:
```suggestion
    // then put it in the context queue so that its response is sent in
```

File: docker/cpu_only/entrypoint.d/12-banner.sh
Comment (dyastremsky) [reply docker/cpu_only/entrypoint.d/12-banner.sh]: I think I tried to over-standardize based on Cliff's comment [here](https://github.com/triton-inference-server/server/pull/6022#discussion_r1256428428). I wanted either all of those shell script files to be executable with shebangs or not executable without shebangs.

Thanks for bringing attention to this. I reverted all the changes so that they remain the same, just updating the executable bit to match the file's presence or absence of a shebang.

File: build.py
Comment (dyastremsky) [reply build.py:2489]: Great idea! Created DLIS-5167 to track.

File: qa/L0_backend_python/python_test.py
Comment (rmccorm4) [inline qa/L0_backend_python/python_test.py:298]: TIL that `(element) == element: T` but `(element,) == element: Tuple[T]` CC @dyastremsky @Tabrizian for the fun fact:
```python
# I thought this would be a tuple of one element, but is a string
>>> x = ("hi")
>>> x
'hi'
>>> type(x)
<class 'str'>      # thought this would be a tuple or set

# This is a tuple
>>> x = ("hi",)
>>> x
('hi',)
>>> type(x)
<class 'tuple'>

# Though this is a list rather than a string, not the same rules as tuple
>>> x = ["hi"]
>>> x
['hi']
>>> type(x)
<class 'list'>
```

File: qa/python_models/init_args/model.py
Comment (rmccorm4) [inline qa/python_models/init_args/model.py:39]: probably unused since the test presumably passes, but might as well fix it:
```suggestion
        "model_repository": os.getenv("TRITON_DIR", "/opt/tritonserver")
```

File: qa/common/sequence_util.py
Comment (rmccorm4) [inline qa/common/sequence_util.py:1]: Still don't totally understand. Did you manually add these shebangs? Or precommit complained about it specifically? Some of these utils probably don't need these nor to be executable - but not super important imo

File: qa/L0_secure_grpc/test.sh
Comment (rmccorm4) [inline qa/L0_secure_grpc/test.sh:76]: These are all wrong and should remain `-passin` - how do we fix this rule?

File: qa/L0_perf_analyzer/test.sh
Comment (rmccorm4) [inline qa/L0_perf_analyzer/test.sh:1003]: wrong, should remain `-passin`, same as L0_secure_grpc

**Review by rmccorm4 — COMMENTED** [review]
Done looking at everything, just a few open comments above before approval with passing CI

File: qa/L0_secure_grpc/test.sh
Comment (dyastremsky) [reply qa/L0_secure_grpc/test.sh]: Thanks for catching these! I thought I already updated the codespell config for this repo to allow `passin`, but perhaps it somehow got lost during a rebasing. The fix to this is to add this line to the pyproject.toml under codespell (done and reverted these changes for the three tests that use this flag):
`# ignore allowed words
ignore-words-list = "passin"`

**Review by rmccorm4 — APPROVED** [review]
Approving pending new passing CI run

File: qa/L0_java_resnet/ResnetTest.java
Comment (rmccorm4) [inline qa/L0_java_resnet/ResnetTest.java:2]: Remove the `(c)` so that this fits on one line. It is failing `L0_copyrights` due to this. Please fix for all 3 files below:
```
+ qa/common/check_copyright.py --year=2023 CITATION.cff CMakeLists.txt CONTRIBUTING.md Dockerfile.QA Dockerfile.sdk Dockerfile.win10.min LICENSE NVIDIA_Deep_Learning_Container_License.pdf README.md TRITON_VERSION Triton-CCLA-v1.pdf build build.py compose.py deploy docker docs pyproject.toml qa src
copyright year is not recognized for qa/L0_java_memory_growth/MemoryGrowthTest.java: // Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights
copyright year is not recognized for qa/L0_java_resnet/ResnetTest.java: // Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights
copyright year is not recognized for qa/L0_java_sequence_batcher/SequenceTest.java: // Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights
```

Comment (dyastremsky) [reply qa/L0_java_resnet/ResnetTest.java]: Thanks for spotting this and recommending a fix. I think I thought it was previously failing and had missed the actual error. Fixed per your recommendation. This now passes in the latest CI.

**Pull request merged**
